plugins/basic/python/tk_maya_basic/plugin_logic.py

```diff
@@ -75,60 +79,134 @@ def login_user():
+        # When the user is not yet authenticated,
+        # pop up the Shotgun login dialog to get the user's credentials,
         # otherwise, get the cached user's credentials.
         user = sgtk.authentication.ShotgunAuthenticator().get_user()
 
     except sgtk.authentication.AuthenticationCancelled:
-        # When the user cancelled the Shotgun login dialog, keep around the displayed login menu.
+        # When the user cancelled the Shotgun login dialog,
+        # keep around the displayed login menu.
         standalone_logger.info("Shotgun login was cancelled by the user.")
         return
 
-    try:
+    # Get rid of the displayed login menu since the engine menu will take over.
+    delete_login_menu()
+
+    # Create a non-interruptable main progress bar (normally in the Help Line).
+    main_progress_bar = pm.ui.MainProgressBar(minValue=0, maxValue=100, interruptable=False)
+    main_progress_bar.beginProgress()
+
+    # Before bootstrapping the engine for the first time around,
+    # the toolkit manager may swap the toolkit core to its latest version.
+    plugin_engine.bootstrap(user,
+                            progress_callback=handle_bootstrap_progress,
+                            completed_callback=handle_bootstrap_completed,
+                            failed_callback=handle_bootstrap_failed)
+
+
+def handle_bootstrap_progress(step_number, message, current_index, maximum_index):
+    """
+    Callback function that reports back on the toolkit and engine bootstrap progress.
+
+    This function is executed in the main thread by the main event loop.
+
+    :param step_number: Current progress step number,
+                        from 1 to ``sgtk.bootstrap.ToolkitManager.MAX_PROGRESS_STEP_NUMBER``.
+    :param message: Progress message to report.
+    :param current_index: Optional current item number being looped over.
+    :param maximum_index: Optional maximum item number being looped over.
+    """
+
+    if maximum_index and maximum_index > 1:
+        message = "%s (%s of %s)" % (message, current_index+1, maximum_index+1)
+
+    message = "Bootstrapping %s: %s" % (manifest.engine_name, message)
+
+    standalone_logger.info(message)
+
+    # Update the main progress bar.
+    progress_value = int((step_number - 1.0) / sgtk.bootstrap.ToolkitManager.MAX_PROGRESS_STEP_NUMBER * 100.0)
+    main_progress_bar.setProgress(progress_value)
+    main_progress_bar.setStatus(message)
+
+    # Force Maya to process its UI events in order to refresh the main progress bar.
+    QtGui.qApp.processEvents()
+
+
+def handle_bootstrap_completed(engine):
+    """
+    Callback function that handles cleanup after successful completion of the bootstrap.
+
+    This function is executed in the main thread by the main event loop.
+
+    :param engine: Launched :class:`sgtk.platform.Engine` instance.
+    """
 
-        try:
-            pm.waitCursor(state=True)
+    # Needed global to re-import the toolkit core.
+    global sgtk
```

Comment:
From https://docs.python.org/2/reference/simple_stmts.html#the-import-statement:

> The first form of import statement binds the module name in the local namespace to the module object, ...

In other words, without the `global`, the `sgtk` name binding at the `plugin_logic` level would not be updated.